sageresearch-app-sdk/src/main/java/org/sagebionetworks/research/sageresearch/reminders/ReminderManager.kt

```diff
@@ -0,0 +1,251 @@
+     */
+    protected val gson = Gson()
+
+    /**
+     * @property reminderAlarmReceiver the receiver to associate with this manager
+     */
+    open val reminderAlarmReceiver: Class<*>
+        get() {
+            return ReminderAlarmReceiver::class.java
+        }
+
+    /**
+     * @property allActivePendingIntents active pending intents must always have the same value for
+     *                                   notifications that are currently scheduled to be properly canceled
+     */
+    protected fun allActivePendingIntents(context: Context): List<PendingIntent> {
+        return allActiveReminders(context).map {
+            pendingIntentForReminder(context, it)
+        }
+    }
+
+    /**
+     * @param context can be app, activity, or service
+     * @return reminders that are currently scheduled to be shown
+     */
+    protected fun allActiveReminders(context: Context): List<Reminder> {
+        return pendingIntentPrefs(context).all.mapNotNull {
+            (it.value as? String)?.let { reminderJson ->
+                return@mapNotNull reminderFromJson(reminderJson)
+            } ?: run {
+                logger.warn("Invalid reminder value with guid ${it.key}")
+            }
+            return@mapNotNull null
+        }
+    }
+
+    /**
+     * Converts a reminder json string to a Reminder
+     * @param reminderJson formed by serializing a Reminder object with Gson
+     * @return reminder formed from reminderJson using Gson
+     */
+    protected fun reminderFromJson(reminderJson: String): Reminder? {
+        try {
+            return gson.fromJson(reminderJson, Reminder::class.java)
+        } catch (e: JsonSyntaxException) {
+            logger.warn(e.localizedMessage)
+            return null
+        }
+    }
+
+    /**
+     * For reminder to be canceled, each reminder must produce the same PendingIntent each time
+     * @param context can be app or activity
+     * @param reminder to be turned into a PendingIntent
+     * @return a pending intent for the reminder data class
+     */
+    protected fun pendingIntentForReminder(context: Context, reminder: Reminder): PendingIntent {
+        val intent = Intent(context, reminderAlarmReceiver)
+        intent.action = reminder.action
+        intent.putExtra(REMINDER_KEY_GUID, reminder.guid)
+        intent.putExtra(REMINDER_KEY_CODE, reminder.code)
+        intent.putExtra(REMINDER_KEY_TITLE, reminder.title)
+        intent.putExtra(REMINDER_KEY_TEXT, reminder.text)
+        intent.data = Uri.parse(reminder.guid)
+        return PendingIntent.getBroadcast(context, reminder.code, intent, PendingIntent.FLAG_ONE_SHOT)
+    }
+
+    /**
+     * @param context can be app or activity
+     * @return a shared preferences object used to store PendingIntents
+     */
+    protected fun pendingIntentPrefs(context: Context): SharedPreferences {
+        return context.getSharedPreferences("ReminderManagerPendingIntents", MODE_PRIVATE)
```

Comment:
Can we make this a static val?

Reply:
I'd rather not.  There's no reason it should stick around for the lifetime of the app.

Reply:
However, I should make it a member variable.